planner/optimize.go

```diff
@@ -342,9 +371,18 @@ func getBindRecord(ctx sessionctx.Context, stmt ast.StmtNode) (*bindinfo.BindRec
 	if globalHandle == nil {
 		return nil, ""
 	}
-	bindRecord = globalHandle.GetBindRecord(hash, normalizedSQL, ctx.GetSessionVars().CurrentDB)
+	bindRecord = globalHandle.GetBindRecord(hash, normalizedSQL, "")
+	// Here is the compatibility code.
+	// There may be some bindings that did not contain the DB name in the original SQL, so we need to use the previous matching method.
 	if bindRecord == nil {
-		bindRecord = globalHandle.GetBindRecord(hash, normalizedSQL, "")
+		stmtNode, normalizedSQL, hash = extractSelectAndNormalizeDigest(stmt, "")
+		if stmtNode == nil {
+			return nil, ""
+		}
+		bindRecord = globalHandle.GetBindRecord(hash, normalizedSQL, ctx.GetSessionVars().CurrentDB)
+		if bindRecord == nil {
+			bindRecord = globalHandle.GetBindRecord(hash, normalizedSQL, "")
+		}
```

Comment:
I believe this approach is not the most effective one for compatibility. i.e, using this method, it is common that user queries would fall into this branch, for each query, we would call one more `Restore` compared with previous implementation, this would impose performance penalty for each query. There may be other alternatives, such as during `Update()` of the bindings, we do `Restore()` once for each binding, or we rewrite the original_sql for all existing bindings in the bootstrap function.

Reply:
In fact, even after we restore the old bindings, compared to the previous implementation, we still have to restore one more time, that is, we need to restore the input SQL.